Added a AsyncTextNowBot class, which allows the user to use the bot with Async.

rohanmodil123 opened the pull request:
Title, and also updated the readme.md with a snippet of code for Async.

rohanmodil123 commented:
The check failure is due to stylistic concerns, but the code runs as expected.

george-lim commented on README.md:
````diff
@@ -55,6 +55,41 @@ with sync_playwright() as api:
+Or, if you need to perform this with `Async`, use the following snippet.
+
+```python
+from textnow_bot import AsyncTextNowBot
+from playwright import async_playwright
+import asyncio
+
+async def send():
+    username = "test@example.com"
+    password = "********"
+    recipient = "123-456-7890"
+    message = "Hello world!"
+
+    async with async_playwright() as api:
+        browser = None
+
+        try:
+            browser = await api.firefox.launch()
+            page = await browser.newPage()
+
+            bot = await AsyncTextNowBot(page, None, username, password)
+            await bot.send_message(recipient, message)
+
+            await browser.close()
+        except Exception as e:
+            if browser:
+                await browser.close()
+            print(e)
+
+            raise
+
+asyncio.run(send())
````
Updated import order and removed `print(e)` so examples are consistent.

```suggestion
import asyncio

from playwright import async_playwright
from textnow_bot import AsyncTextNowBot


async def send():
    username = "test@example.com"
    password = "********"
    recipient = "123-456-7890"
    message = "Hello world!"

    async with async_playwright() as api:
        browser = None

        try:
            browser = await api.firefox.launch()
            page = await browser.newPage()

            bot = await AsyncTextNowBot(page, None, username, password)
            await bot.send_message(recipient, message)

            await browser.close()
        except Exception:
            if browser:
                await browser.close()

            raise

asyncio.run(send())
```

george-lim commented on textnow_bot.py:
```diff
@@ -44,3 +46,56 @@ def send_message(self, recipient, message):
+    async def __init__(self, page, cookies=None, username=None, password=None):
+        self.page = page
+
+        if cookies:
+            logging.info("Async logging in with cookies...")
+            await page.context.addCookies(cookies)
+            await page.goto(f"{TEXTNOW_URL}/login", waitUntil="networkidle")
+        elif username and password:
+            logging.info("Async logging in with account info...")
+            await page.context.clearCookies()
+
+            await page.goto(f"{TEXTNOW_URL}/login")
+            await page.type("#txt-username", username)
+            await page.type("#txt-password", password)
+            await page.click("#btn-login")
+            await page.waitForNavigation()
+
+            await page.context.addCookies([PERMISSION_COOKIE])
+        else:
+            raise Exception("missing authentication info")
+
+        if "/messaging" not in page.url:
+            # Added more information to error message
+            raise Exception("authentication failed (did not reach textnow.com/messaging)")
+
+    async def get_cookies(self):
+        return await self.page.context.cookies(TEXTNOW_URL)
+
+    async def send_message(self, recipient, message):
+        logging.info("Async sending message...")
+
+        await self.page.goto(f"{TEXTNOW_URL}/messaging")
+        await self.page.click("#newText")
+
+        await self.page.type(".newConversationTextField", recipient)
+        await self.page.press(".newConversationTextField", "Enter")
+
+        await self.page.type("#text-input", message)
+        await self.page.press("#text-input", "Enter")
+        await self.page.waitForTimeout(500)
```
Address linter changes.

Finally, can we add a comment that explains what `AsyncMeta` does? It is not immediately obvious what it is doing, but this could be due to my of experience with asynchronous programming in Python. I assume that it is meant for debugging purposes? If so, does it make sense to have it in the module?

```suggestion
class AsyncMeta(type):
    async def __call__(cls, *args, **kwargs):
        obb = object.__new__(cls)
        fn = obb.__init__(*args, **kwargs)

        if inspect.isawaitable(fn):
            await fn

        return obb


class AsyncTextNowBot(metaclass=AsyncMeta):
    async def __init__(self, page, cookies=None, username=None, password=None):
        self.page = page

        if cookies:
            logging.info("Logging in with cookies...")
            await page.context.addCookies(cookies)
            await page.goto(f"{TEXTNOW_URL}/login", waitUntil="networkidle")
        elif username and password:
            logging.info("Logging in with account info...")
            await page.context.clearCookies()

            await page.goto(f"{TEXTNOW_URL}/login")
            await page.type("#txt-username", username)
            await page.type("#txt-password", password)
            await page.click("#btn-login")
            await page.waitForNavigation()

            await page.context.addCookies([PERMISSION_COOKIE])
        else:
            raise Exception("missing authentication info")

        if "/messaging" not in page.url:
            raise Exception(
                "authentication failed: did not reach textnow.com/messaging"
            )

    async def get_cookies(self):
        return await self.page.context.cookies(TEXTNOW_URL)

    async def send_message(self, recipient, message):
        logging.info("Sending message...")

        await self.page.goto(f"{TEXTNOW_URL}/messaging")
        await self.page.click("#newText")

        await self.page.type(".newConversationTextField", recipient)
        await self.page.press(".newConversationTextField", "Enter")

        await self.page.type("#text-input", message)
        await self.page.press("#text-input", "Enter")
        await self.page.waitForTimeout(500)

```

george-lim commented on textnow_bot.py:
```diff
@@ -1,4 +1,6 @@
 import logging
+import inspect
+import asyncio
```
Remove unused `asyncio` import, and also reorder imports for linters.

```suggestion
import inspect
import logging
```

george-lim commented on textnow_bot.py:
```diff
@@ -27,7 +29,7 @@ def __init__(self, page, cookies=None, username=None, password=None):
 
         if "/messaging" not in page.url:
-            raise Exception("authentication failed")
+            raise Exception("authentication failed (did not reach textnow.com/messaging")
```
Personally, I prefer to use colons to delimit exception descriptions from exception messages.

```suggestion
            raise Exception(
                "authentication failed: did not reach textnow.com/messaging"
            )
```

george-lim reviewed (commented):
I have provided suggestions which address linter concerns.

The code looks fine, but I would like to understand what `AsyncMeta` is doing, and whether it is necessary to have in the module.

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +47,56 @@ def send_message(self, recipient, message):
+
+# Code for Async Below
+
+class AsyncMeta(type):
```
```suggestion
class AsyncMeta(type): # Allows the __init__ function of AsyncTextNowBot() to be async.
```

`AsyncMeta` is a class I created to get around the fact that normally, the `__init__` function of a class cannot be asynchronous in Python. 

This issue could also have been solved by creating a `log_in` function, and transferring all the code out of the `__init__` function into the new `log_in` function, but this would mean that instead of using the following code to send a message:

```python
bot = await AsyncTextNowBot(page, None, username, password)
await bot.send_message(recipient, message)
```

You would have to do:

```python
bot = await AsyncTextNowBot(page, None, username, password)
await bot.log_in()
await bot.send_message(recipient, message)
```

Alternatively, you could have the `__init__` function include `asyncio.run(log_in())`, which would avoid the extra command for the user. 

However, I thought that for consistency it would make more sense for each of the two classes (`TextNowBot` and `AsyncTextNowBot`) to have the same methods, so I decided on using the metaclass.


Thanks for addressing the linter concerns.

george-lim replied:
I see, that makes a lot of sense. To give you some context, I specifically added login logic to `__init__` so that functions like `send_message` have the invariant that the user has logged in. It simplifies the need to keep track of states (whether the user is logged in or not, etc.)

I'm not a fan of `AsyncMeta`, because it looks hack-y to me. I feel like I would prefer to just change the API in this case to add a `log_in` method so that both sync_api and async_api have the same experience, and we can do `async log_in()` like you suggested. Overall, it's a cleaner solution in my opinion, but I don't have any strong opinions about that. What do you think?

rohanmodil123 replied:
> I feel like I would prefer to just change the API in this case to add a `log_in` method ...

Yeah, that sounds fine to me. I'll put that together.

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +47,56 @@ def send_message(self, recipient, message):
+        if inspect.isawaitable(fn):
+            await fn
+        return obb
+
```
```suggestion


```

E302: Expected 2 Blank Lines, Found 1 (Linter)

george-lim replied:
Feel free to commit this 👍

george-lim commented:
FYI, an easier way to deal with linter is to run `docker run -e RUN_LOCAL=true -v $PWD:/tmp/lint github/super-linter` in the repo, if you have docker. This will run the super-linter locally. If you don't want to do it, don't worry about it. Truthfully I just have the linter set-up because I want to be strict on my own projects.

rohanmodil123 commented:
I don't actually have docker, and I have never actually used one of these GitHub linters before, which brings me to my next point, this error:

```
 2021-02-01 20:55:33 [ERROR]   Found errors in [black] linter!
2021-02-01 20:55:33 [ERROR]   Error code: 1. Command output:
------
--- textnow_bot.py	2021-02-01 20:55:24.705436 +0000
+++ textnow_bot.py	2021-02-01 20:55:33.952650 +0000
@@ -49,17 +49,19 @@
         self.page.waitForTimeout(500)
 
 
 # Code for Async Below
 
+
 class AsyncMeta(type):
     async def __call__(cls, *args, **kwargs):
         obb = object.__new__(cls)
         fn = obb.__init__(*args, **kwargs)
         if inspect.isawaitable(fn):
             await fn
         return obb
+
 
 class AsyncTextNowBot(metaclass=AsyncMeta):
     async def __init__(self, page, cookies=None, username=None, ***
         self.page = page
 
@@ -81,11 +83,13 @@
         else:
             raise Exception("missing authentication info")
 
         if "/messaging" not in page.url:
             # Added more information to error message
-            raise Exception("authentication failed (did not reach textnow.com/messaging)")
+            raise Exception(
+                "authentication failed (did not reach textnow.com/messaging)"
+            )
 
     async def get_cookies(self):
         return await self.page.context.cookies(TEXTNOW_URL)
 
     async def send_message(self, recipient, message):
would reformat textnow_bot.py
Oh no! 💥 💔 💥
1 file would be reformatted.
```

I can't tell what line it is erroring on, or what the error is...
I have run the code on my own system and it runs without error, so I am unsure of what the linter is pointing out here. 
Of course, I have not used this linter before this, so it could be something trivial that I am missing.

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +47,58 @@ def send_message(self, recipient, message):
+        return obb
+
+
+
```
```suggestion

```

Somehow I added 3 lines? Whoops, taking away one.

george-lim commented:
The error means that you should remove the lines that are prefixed with `-` and replace those lines with the lines that are prefixed with `+`. Don't worry about the linter though. Just make the changes to the API and re-request review from me when you are done. I'll address the linter changes again.

rohanmodil123 commented:
Sorry for the delay. Removed `AsyncMeta` and replaced it with a new `log_in` command, which is automatically run with `__init__`.

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -1,4 +1,7 @@
+import asyncio
+import inspect
 import logging
```
```suggestion
```

rohanmodil123 replied:
Unused import

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -1,4 +1,7 @@
+import asyncio
+import inspect
```
Unused import 
```suggestion
```

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+
+class AsyncTextNowBot():
+    def __init__(self, page, cookies=None, username=None, password=None):
+        #self.log_in(page, cookies, username, password)
```
```suggestion
        # self.log_in(page, cookies, username, password)
```
E265 block comment should start with '# '

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+            await page.context.clearCookies()
+            # Unneccesary: await page.goto(f"{TEXTNOW_URL}/login")
+
+            # This sleep is sometimes neccessary for it to work. Sometimes it works without it. 
```
```suggestion
            # This sleep is sometimes neccessary for it to work. Sometimes it works without it.
```
W291 trailing whitespace

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+            # Unneccesary: await page.goto(f"{TEXTNOW_URL}/login")
+
+            # This sleep is sometimes neccessary for it to work. Sometimes it works without it. 
+            # Also, it sometimes does not work if it is a non-blocking sleep, it has to be a blocking sleep. The wonders of async.
```
```suggestion
            # Sometimes does not work if it is a non-blocking sleep, it has to be a blocking sleep. The wonders of async.
```

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+    async def get_cookies(self):
+        return await self.page.context.cookies(TEXTNOW_URL)
+
+    async def send_message(self, recipient, message):    
```
```suggestion
    async def send_message(self, recipient, message):
```
W291 trailing whitespace

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+            # Unneccesary: await page.goto(f"{TEXTNOW_URL}/login")
+
+            # This sleep is sometimes neccessary for it to work. Sometimes it works without it.
+            # Sometimes does not work if it is a non-blocking sleep, it has to be a blocking sleep. The wonders of async.
```
```suggestion
            # Sometimes does not work with a non-blocking sleep, it has to be a blocking sleep. The wonders of async.
```
E501 line too long (121 > 120 characters)

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+
+
+# Code for Async Below
+
```
```suggestion


```

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+
+# Code for Async Below
+
+class AsyncTextNowBot():
```
```suggestion

class AsyncTextNowBot:
```

linter

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,57 @@ def send_message(self, recipient, message):
+            raise Exception("missing authentication info")
+
+        if "/messaging" not in page.url:
+            raise Exception("authentication failed: did not reach textnow.com/messaging")
```
```suggestion
            raise Exception(
            "authentication failed: did not reach textnow.com/messaging"
            )        
```
linter

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+
+        if "/messaging" not in page.url:
+            raise Exception(
+            "authentication failed: did not reach textnow.com/messaging"
```
```suggestion
                "authentication failed: did not reach textnow.com/messaging"
```

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+        if "/messaging" not in page.url:
+            raise Exception(
+            "authentication failed: did not reach textnow.com/messaging"
+            )        
```
```suggestion
            )
```

rohanmodil123 commented:
Adapted the code to fit the linter, and made the change you requested regarding `AsyncMeta`

Removed `AsyncMeta` and replaced it with a new `log_in` command, which is automatically run with `__init__`.

george-lim commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+
+class AsyncTextNowBot:
+    def __init__(self, page, cookies=None, username=None, password=None):
+        # self.log_in(page, cookies, username, password)
+        print(page)
```
Let's remove debug statements in the module please 🙏

rohanmodil123 replied:
Whoops, fixed.

george-lim commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+        elif username and password:
+            logging.info("Async logging in with account info...")
+            await page.context.clearCookies()
+            # Unneccesary: await page.goto(f"{TEXTNOW_URL}/login")
```
why is this unnecessary? page does not have an initial URL defined otherwise. We are assuming that the page that is passed into the class in `__init__` is empty. The user should not be responsible for knowing the URL.

rohanmodil123 replied:
This is unnecessary as async will start to `send_message`. This will take the page to `textnow.com/messaging`. Since the user is not logged in, it will redirect to `textnow.com/login`, and it will proceed to login. 

I commented it out as that line causes errors for some reason, and I figured considering it works without it, I need not attempt to fix that.

rohanmodil123 replied:
> This is unnecessary as async will start to send_message. This will take the page to textnow.com/messaging. Since the user is not logged in, it will redirect to textnow.com/login, and it will proceed to login.

The code works using the example code given in the readme, and the user does not need to know the login URL.

george-lim commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+            await page.context.clearCookies()
+            # Unneccesary: await page.goto(f"{TEXTNOW_URL}/login")
+
+            # This sleep is sometimes neccessary for it to work. Sometimes it works without it.
+            # Sometimes does not work with a non-blocking sleep, it has to be a blocking sleep. The wonders of async.
+            time.sleep(1)
```
Appreciate the comment, though that is strange 👍

rohanmodil123 replied:
Yeah, this confused me. The could just would not work, so I decided to just add sleep statements to see where things would go wrong, as the error messages were giving me errors on lines that weren't even being reached, and I didn't feel like debugging. After the sleep statement was added, it miraculously worked, so I played around with it and found that to work consistently it needed at least 1 second of blocking sleep. I'm guessing it's probably something with playwright loading, although I'm not sure.

george-lim reviewed (commented):
Almost there! Just have some questions and minor feedback :)

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+
+class AsyncTextNowBot:
+    def __init__(self, page, cookies=None, username=None, password=None):
+        # self.log_in(page, cookies, username, password)
```
```suggestion
```

rohanmodil123 commented on textnow_bot.py:
```diff
@@ -44,3 +48,60 @@ def send_message(self, recipient, message):
+class AsyncTextNowBot:
+    def __init__(self, page, cookies=None, username=None, password=None):
+        # self.log_in(page, cookies, username, password)
+        print(page)
```
```suggestion
```
Whoops.

george-lim commented:
By the way, please press the re-review button so it notifies me that you're done with your changes. You can find it by pressing the cycle button on the right. I happened to check this PR today, so I assumed that you were done. If you press the button, then I know that you are ready for me to review again.

<img width="317" alt="Screen Shot 2021-02-05 at 1 24 28 PM" src="https://user-images.githubusercontent.com/21700768/107073704-af234d80-67b5-11eb-9e90-05ede63728b5.png">

rohanmodil123 commented:
I have the removed the debugging code. 

Also, I have responded to your message as to why `await page.goto(f"{TEXTNOW_URL}/login")` is unnecessary.

george-lim commented on textnow_bot.py:
```diff
@@ -44,3 +48,58 @@ def send_message(self, recipient, message):
+        else:
+            raise Exception("missing authentication info")
+
+        if "/messaging" not in page.url:
+            raise Exception(
+                "authentication failed: did not reach textnow.com/messaging"
+            )
```
The `await page.goto(f"{TEXTNOW_URL}/login")` line is necessary because of this code, which checks logging in leads to `/messaging` endpoint.

For reference: checking whether /messaging exists in the page URL is the method I use to check if the user is logged in. If you enter an invalid username and password, the user is stuck on the login URL, so this check will fail. Similarly, if cookies are invalidated or expire, then this check will fail because the user will be redirected to the Textnow logout page.

The await goto command ensures that this check is run, because it runs the login page necessary for this check to work.

rohanmodil123 replied:
By line #70, Playwright will be on textnow.com/login, as it will have been redirected from textnow.com/messaging.

By line #79, if all credentials are correct, playwright will be back on textnow.com/messaging.

If the credentials are not correct or the login has somehow failed, line #79 will still be on textnow.com/login. page.url will confirm this (I just tried it). 

The check does not fail if the correct credentials are given. (Also tested this)

The removal of the code on line #71 - the one that does page.goto, does not change the function of any of the remaining code. It will login, redirect to textnow.com/messaging, and then pass your check.

If the credentials were wrong, the login will fail, it will not redirect to textnow.com/messaging, and the code will error as a result of ` page.waitForNavigation()`, as it will wait forever to be redirected, which will not happen, and it will timeout. This is not an issue that was added/changed by the removal of `page.goto` 

 > there is no initial URL that is loaded.

This is false.

rohanmodil123 replied:
The issue with `page.waitForNavigation()` could be resolved by catching a `TimeoutError` and instead throwing your custom error message, but once again it has nothing to do with removing `page.goto("textnow.com/login")`

george-lim commented:
Alright, so spent the morning learning asyncio, and I now fully understand the problem. What you are doing when you do `ensure_future` is you are running the log_in code in a background thread. You need to have a URL to log_in. However, because you are using the same `self.page` member across both threads, you used `time.sleep` to basically wait until the send_message thread loads the URL, then the log_in thread has a URL to login.

This solution has many race-conditions. A better solution is to enforce users to call `log_in` before calling `send_message`. This way, we don't need to worry about race conditions. I'm going to close your PR because I think I have a clean solution that addresses this feature. I will make a PR and tag you as a reviewer.
